### PR TITLE
Add GenericAudio Backend, make SDL mixer optional

### LINKED ISSUE #968: Support playback through SDL2 without Mixer dependency
Should be feasible when all formats are supported by our decoders ^^

via https://gist.github.com/armornick/3447121

``` cpp
#include <SDL2/SDL.h>

#define MUS_PATH "Roland-GR-1-Trumpet-C5.wav"

// prototype for our audio callback
// see the implementation for more information
void my_audio_callback(void *userdata, Uint8 *stream, int len);

// variable declarations
static Uint8 *audio_pos; // global pointer to the audio buffer to be played
static Uint32 audio_len; // remaining length of the sample we have to play


/*
** PLAYING A SOUND IS MUCH MORE COMPLICATED THAN IT SHOULD BE
*/
int main(int argc, char* argv[]){

    // Initialize SDL.
    if (SDL_Init(SDL_INIT_AUDIO) < 0)
            return 1;

    // local variables
    static Uint32 wav_length; // length of our sample
    static Uint8 *wav_buffer; // buffer containing our audio file
    static SDL_AudioSpec wav_spec; // the specs of our piece of music


    /* Load the WAV */
    // the specs, length and buffer of our wav are filled
    if( SDL_LoadWAV(MUS_PATH, &wav_spec, &wav_buffer, &wav_length) == NULL ){
      return 1;
    }
    // set the callback function
    wav_spec.callback = my_audio_callback;
    wav_spec.userdata = NULL;
    // set our global static variables
    audio_pos = wav_buffer; // copy sound buffer
    audio_len = wav_length; // copy file length

    /* Open the audio device */
    if ( SDL_OpenAudio(&wav_spec, NULL) < 0 ){
      fprintf(stderr, "Couldn't open audio: %s\n", SDL_GetError());
      exit(-1);
    }

    /* Start playing */
    SDL_PauseAudio(0);

    // wait until we're don't playing
    while ( audio_len > 0 ) {
        SDL_Delay(100); 
    }

    // shut everything down
    SDL_CloseAudio();
    SDL_FreeWAV(wav_buffer);

}

// audio callback function
// here you have to copy the data of your audio buffer into the
// requesting audio buffer (stream)
// you should only copy as much as the requested length (len)
void my_audio_callback(void *userdata, Uint8 *stream, int len) {

    if (audio_len ==0)
        return;

    len = ( len > audio_len ? audio_len : len );
    //SDL_memcpy (stream, audio_pos, len);                  // simply copy from one buffer into the other
    SDL_MixAudio(stream, audio_pos, len, SDL_MIX_MAXVOLUME);// mix from one buffer into another

    audio_pos += len;
    audio_len -= len;
}
```

### PR DESCRIPTION
**Pull request by Ghabry**
This uses work of @ChristianBreitwieser who wrote a generic audio mixer (originally for libretro, but I took it and made is more suitable for general purpose).

All audio backends (Except 3DS - lack of CPU power - and SDL Mixer - obvious reasons) use the GenericAudio backend now which simplified the code a lot.

Removed SDL Mixer dependency from Wii.
Android still depends on Mixer due to lack of WildMidi (buildsystem upgrade required)

New arguments for configure:
```
--without-audio
--without-sdlmixer
```

Fixes #968

### CODE REVIEW

File: src/audio_decoder.cpp [hunk 1]
```diff
@@ -143,6 +143,7 @@ std::unique_ptr<AudioDecoder> AudioDecoder::Create(FILE* file, const std::string
 	if (!strncmp(magic, "MThd", 4)) {
 #ifdef HAVE_WILDMIDI
 		static bool wildmidi_works = true;
+		printf("MIDI\n");
```
Comment (carstene1ns) [inline src/audio_decoder.cpp]: debug leftover?

**Comment by Ghabry** [comment]
Reminder for me:
- Check for SUPPORT_AUDIO in all Makefiles
- when none of the linked audio libraries supports the format <- supportS, typo. linked - provided?

**Comment by Ghabry** [comment]
I think it would make sense to enable Wildmidi and FmMidi by default when SDL Mixer is not found (did this in CMake). opinion?

**Comment by Ghabry** [comment]
WildMidi doesn't like my "/etc/timidity++/timidity.cfg" on Arch, silence

**Comment by Ghabry** [comment]
Submitted a patch to WildMidi to fix playback on Arch: https://github.com/Mindwerks/wildmidi/pull/166

File: src/audio_generic.cpp [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+				}
+
+				if (!channel_active) {
+					//first channel
+					mixer_buffer.data()[ii * output_channels] = vall;
+					if (channels > 1) {
+						mixer_buffer.data()[ii * output_channels + 1] = valr;
+					} else {
+						mixer_buffer.data()[ii * output_channels + 1] = mixer_buffer.data()[ii *
+																							output_channels];
```
Comment (scurest) [inline src/audio_generic.cpp]: Weird indent (and below too).

**Comment by scurest** [comment]
Also, in `audio_generic.cpp` there are a lot of calls like `vec.data()[index]`. Excepting the ones where you cast the `data` pointer, those can be `vec[index]`, right?

**Comment by Ghabry** [comment]
You are right about ```data()```, that's not required.

In general the formating of GenericAudio is not the best, will check.

More notes for me:
Move the "static" variables in Decode in the private part of the class.
Try to reduce amount of Mutex locks in BGM_* and SE_* API.

**Comment by Ghabry** [comment]
About wildmidi: They backport the fix to a new version (0.3.12) unfortunately Arch won't profit from this because they package 0.4 ( this versioning scheme ^^)

File: CMakeLists.txt [hunk 3]
```diff
@@ -258,15 +260,78 @@ if(Freetype_FOUND)
...
+	option(PLAYER_WITH_WILDMIDI "Play MIDI audio with wildmidi." ON)
+	find_lib("WildMidi" PLAYER_WITH_WILDMIDI)
+	if(WildMidi_FOUND)
+		add_definitions(-DHAVE_WILDMIDI)
+	endif()
+	
+	# Provide fmmidi options
+	option(PLAYER_ENABLE_FMMIDI "Enable internal MIDI sequencer. Will be used when external MIDI library fails." ON)
+	if(${PLAYER_ENABLE_FMMIDI} MATCHES "ON")
+		add_definitions(-DWANT_FMMIDI)
```
Comment (carstene1ns) [inline CMakeLists.txt]: Does this pass the check in `system.h`? [1/2]

File: builds/psvita/Makefile [hunk 4]
```diff
@@ -28,7 +28,7 @@ CC      = $(PREFIX)-gcc
 CXX      = $(PREFIX)-g++
 CFLAGS  = -Wl,-q -g -O3 -DPSP2 -DHAVE_LIBSPEEXDSP \
 		-DWANT_FASTWAV -DSUPPORT_AUDIO -DHAVE_LIBSNDFILE \
-		-DHAVE_OGGVORBIS -DHAVE_MPG123 -DWANT_FMMIDI=1 \
+		-DHAVE_OGGVORBIS -DHAVE_MPG123 -DWANT_FMMIDI \
```
Comment (carstene1ns) [inline builds/psvita/Makefile]: Does this pass the check in `system.h`? [2/2]

Comment (Ghabry) [inline builds/psvita/Makefile]: Yes, because defines default to 1

**Comment by Ghabry** [comment]
```
General question: Does the 3DS change not to use CSND anymore and removing libkhax earlier still need a target app for permissions,etc? If not, we should remove the easyrpg-player.xml from builds/3ds folder...
```
CSND and khax are gone. The easyrpg-player.xml is for selecting the youtube-channel as a target. This is still required for non-CIA due to size limits of the executable memory.

**Comment by Ghabry** [comment]
I have a one-line-change to change the newlib-heapsize on the 3DS, shall I add this, too?

File: src/audio_3ds.cpp [hunk 5]
```diff
@@ -16,514 +16,328 @@
...
-	}
-
-	// Opening and decoding the file
-	DecodedMusic* myFile = (DecodedMusic*)malloc(sizeof(DecodedMusic));
-	int res = DecodeMusic(file, myFile);
-	if (res < 0){
-		free(myFile);
+	FILE* filehandle = FileFinder::fopenUTF8(file, "rb");
+	if (!filehandle) {
+		Output::Warning("Audio not readable: %s", file.c_str());
```
Comment (carstene1ns) [inline src/audio_3ds.cpp]: "Music" ?

File: src/audio_3ds.cpp [hunk 6]
```diff
@@ -16,514 +16,328 @@
...
+		bgm_decoder->GetFormat(frequency, format, channels);
+		bgm_decoder->SetFade(0,volume,fadein);
+		bgm_decoder->SetLooping(true);
+
+		if (!set_channel_format(bgm_channel, format, channels, out_format)) {
+			bgm_decoder->SetFormat(frequency, out_format, channels);
+		}
+		ndspChnSetRate(bgm_channel, frequency);
+
+		Output::Debug("Audio started: %s, samplerate: %u, pitch: %u", file.c_str(),frequency,pitch);
```
Comment (carstene1ns) [inline src/audio_3ds.cpp]: Debug leftover?

**Comment by carstene1ns** [comment]
I get crashes when running player without sdl_mixer:
```
Debug: EasyRPG Player 0.5.0 started
Debug: ============================
Debug: CLI: --project-path rpg/ --test-play --window --show-fps 
[New Thread 0x7ffff120d700 (LWP 12373)]
[New Thread 0x7fffe7359700 (LWP 12374)]
[New Thread 0x7fffe6953700 (LWP 12375)]
[New Thread 0x7fffe6152700 (LWP 12376)]
[New Thread 0x7fffe5951700 (LWP 12377)]
[New Thread 0x7fffe5150700 (LWP 12378)]
[New Thread 0x7fffcfaeb700 (LWP 12379)]
[New Thread 0x7fffe40c7700 (LWP 12380)]
Debug: Detected encoding: ibm-5348_P100-1997
Debug: Using rpg//Velsarbor as Game and Save directory
Debug: Loading game Velsarbor
Debug: Using RPG2k Interpreter
Debug: Non-official Harmony.dll found, skipping MP3 test
Debug: RPG2k < v1.50 / RPG2k3 < v1.05 detected
Debug: Engine configured as: 2k=1 2k3=0 2k3Legacy=0 MajorUpdated=0 2k3E=0
Debug: wine prefix not found: "/home/carsten/.wine"
Debug: wine prefix not found: "/home/carsten/.wine"
Debug: wine prefix not found: "/home/carsten/.wine"
Debug: wine prefix not found: "/home/carsten/.wine"
Warning: RTP not found. This may create missing file errors.
Install RTP files or check they are installed fine.
If this game really does not require RTP, then add
FullPackageFlag=1 line to the RPG_RT.ini game file.
Debug: Actor 5: Removing invalid item 17 (of type 0) from equipment slot 1 (needs type 1)
Debug: Actor 8: Removing invalid item 17 (of type 0) from equipment slot 1 (needs type 1)
Debug: Actor 8: Removing invalid item 53 (of type 1) from equipment slot 3 (needs type 3)
Debug: Actor 9: Removing invalid item 17 (of type 0) from equipment slot 1 (needs type 1)
Debug: Actor 9: Removing invalid item 53 (of type 1) from equipment slot 3 (needs type 3)
Debug: WildMidi: Using wildmidi.cfg as configuration file...
Debug: Starting new game
Debug: Loading Map Map0016.lmu
Debug: Tree: ID-Eingabe < Intro1-ANFANG

Thread 9 "SDLAudioDev1" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe40c7700 (LWP 12380)]
GenericAudio::Decode (this=0xfbfc30, 
    output_buffer=0xf99510 "L\002L\002\002\002\002\002\313\001\313\001\003\002\003\002>\002>\002\371\001\371\001[\001[\001\v\001\v\001Z\001Z\001\361\001\361\001C\002C\002\022\002\022\002\201\001\201", <incomplete sequence \354>, 
    buffer_length=3760) at src/audio_generic.cpp:381
381						mixer_buffer[ii * output_format.channels] = vall;
(gdb) bt f
#0  GenericAudio::Decode (this=0xfbfc30, 
    output_buffer=0xf99510 "L\002L\002\002\002\002\002\313\001\313\001\003\002\003\002>\002>\002\371\001\371\001[\001[\001\v\001\v\001Z\001Z\001\361\001\361\001C\002C\002\022\002\022\002\201\001\201", <incomplete sequence \354>, 
    buffer_length=3760) at src/audio_generic.cpp:381
        vall = 0
        valr = 0
        ii = 15564
        read_bytes = 15525120
        channels = 2
        volume = 0
        samplesize = 2
        frequency = 44100
        sampleformat = AudioDecoder::Format::S16
        is_bgm_channel = <optimized out>
        channel_used = <optimized out>
        i = 1
        channel_active = false
        total_volume = 0
        samples_per_frame = 940
#1  0x00007ffff6f009a2 in ?? () from /usr/lib/libSDL2-2.0.so.0
No symbol table info available.
#2  0x00007ffff6f6343c in ?? () from /usr/lib/libSDL2-2.0.so.0
No symbol table info available.
#3  0x00007ffff6fb48e9 in ?? () from /usr/lib/libSDL2-2.0.so.0
No symbol table info available.
#4  0x00007ffff29422e7 in start_thread () from /usr/lib/libpthread.so.0
No symbol table info available.
#5  0x00007ffff519554f in clone () from /usr/lib/libc.so.6
No symbol table info available.
(gdb) q
A debugging session is active.

	Inferior 1 [process 12368] will be killed.

Quit anyway? (y or n) y
```

**Comment by Ghabry** [comment]
One problem is when volume is 0, then "read_bytes" is not initialized. But even after fixing this ASAN still reports a crash.
Interestingly this doesn't crash under Windows, not even with application verifier enabled o.O

**Comment by Ghabry** [comment]
Had to remove the 3ds heapsize change. Crashes on real hardware. So we have to instead reduce liblcf mem footprint ;)

**Comment by carstene1ns** [comment]
Merging this now. Will test the Wii version later on hardware...

**Pull request merged**